pkg/streamingpromql/operators/aggregation.go

```diff
@@ -152,61 +156,99 @@ func (a *Aggregation) NextSeries(ctx context.Context) (types.InstantVectorSeries
 		thisSeriesGroup := a.remainingInnerSeriesToGroup[0]
 		a.remainingInnerSeriesToGroup = a.remainingInnerSeriesToGroup[1:]
 
-		if thisSeriesGroup.sums == nil {
+		if len(s.Floats) > 0 && thisSeriesGroup.floatSums == nil {
 			// First series for this group, populate it.
 
-			thisSeriesGroup.sums, err = a.Pool.GetFloatSlice(steps)
+			thisSeriesGroup.floatSums, err = a.Pool.GetFloatSlice(steps)
 			if err != nil {
 				return types.InstantVectorSeriesData{}, err
 			}
 
-			thisSeriesGroup.present, err = a.Pool.GetBoolSlice(steps)
+			thisSeriesGroup.floatPresent, err = a.Pool.GetBoolSlice(steps)
 			if err != nil {
 				return types.InstantVectorSeriesData{}, err
 			}
+			thisSeriesGroup.floatSums = thisSeriesGroup.floatSums[:steps]
+			thisSeriesGroup.floatPresent = thisSeriesGroup.floatPresent[:steps]
+		}
 
-			thisSeriesGroup.sums = thisSeriesGroup.sums[:steps]
-			thisSeriesGroup.present = thisSeriesGroup.present[:steps]
+		if len(s.Histograms) > 0 && thisSeriesGroup.histogramValues == nil {
+			// First series for this group, populate it.
+			thisSeriesGroup.histogramValues, err = a.Pool.GetHistogramPointerSlice(steps)
+			if err != nil {
+				return types.InstantVectorSeriesData{}, err
+			}
+			thisSeriesGroup.histogramValues = thisSeriesGroup.histogramValues[:steps]
 		}
 
 		for _, p := range s.Floats {
 			idx := (p.T - start) / interval
-			thisSeriesGroup.sums[idx] += p.F
-			thisSeriesGroup.present[idx] = true
+			thisSeriesGroup.floatSums[idx] += p.F
+			if !thisSeriesGroup.floatPresent[idx] {
+				thisSeriesGroup.floatPointCount += 1
+			}
```

Comment:
It'd be interesting to compare benchmarks with and without this change - I suspect iterating over `floatPresent` once just before we produce the group is cheaper than this check on every single input sample, but let's see what the data says.

Reply:
The old code still has a check:
```
	pointCount := 0
	for _, p := range thisGroup.present {
		if p {
			pointCount++
```
So not sure if this would be any worse.

Nonetheless, I've opened a separate PR for this here along with a benchmark: https://github.com/grafana/mimir/pull/8442

I'm going to resolve this and move the conversation over there. If we merge that one, I'll rebase this.